File: src/Services/TalyerStudio.JobOrder/TalyerStudio.JobOrder.Application/Services/JobOrderService.cs
using TalyerStudio.JobOrder.Application.DTOs;
using TalyerStudio.JobOrder.Application.Interfaces;

namespace TalyerStudio.JobOrder.Application.Services;

public class JobOrderService : IJobOrderService
{
    private readonly IJobOrderRepository _repository;

    public JobOrderService(IJobOrderRepository repository)
    {
        _repository = repository;
    }

    public async Task<JobOrderDto?> GetByIdAsync(Guid id)
    {
        var jobOrder = await _repository.GetByIdAsync(id);
        return jobOrder == null ? null : MapToDto(jobOrder);
    }

    public async Task<List<JobOrderDto>> GetAllAsync(Guid tenantId, string? status = null, int skip = 0, int take = 50)
    {
        var jobOrders = await _repository.GetAllAsync(tenantId, status, skip, take);
        return jobOrders.Select(MapToDto).ToList();
    }

    public async Task<JobOrderDto> CreateAsync(CreateJobOrderDto dto, Guid tenantId)
    {
        var jobOrder = await _repository.CreateAsync(dto, tenantId);
        return MapToDto(jobOrder);
    }

    public async Task<bool> UpdateStatusAsync(Guid id, UpdateJobOrderStatusDto dto)
    {
        return await _repository.UpdateStatusAsync(id, dto);
    }

    public async Task<bool> AssignMechanicsAsync(Guid id, AssignMechanicDto dto)
    {
        return await _repository.AssignMechanicsAsync(id, dto);
    }

    public async Task<bool> DeleteAsync(Guid id)
    {
        return await _repository.DeleteAsync(id);
    }

    public async Task<List<JobOrderDto>> GetByCustomerIdAsync(Guid customerId)
    {
        var jobOrders = await _repository.GetByCustomerIdAsync(customerId);
        return jobOrders.Select(MapToDto).ToList();
    }

    public async Task<List<JobOrderDto>> GetByVehicleIdAsync(Guid vehicleId)
    {
        var jobOrders = await _repository.GetByVehicleIdAsync(vehicleId);
        return jobOrders.Select(MapToDto).ToList();
    }

    private JobOrderDto MapToDto(Domain.Entities.JobOrder jobOrder)
    {
        return new JobOrderDto
        {
            Id = jobOrder.Id,
            JobOrderNumber = jobOrder.JobOrderNumber,
            TenantId = jobOrder.TenantId,
            BranchId = jobOrder.BranchId,
            CustomerId = jobOrder.CustomerId,
            VehicleId = jobOrder.VehicleId,
            Status = jobOrder.Status.ToString(),
            Priority = jobOrder.Priority.ToString(),
            OdometerReading = jobOrder.OdometerReading,
            CustomerComplaints = jobOrder.CustomerComplaints,
            InspectionNotes = jobOrder.InspectionNotes,
            BeforePhotos = jobOrder.BeforePhotos,
            AfterPhotos = jobOrder.AfterPhotos,
            AssignedMechanicIds = jobOrder.AssignedMechanicIds,
            StartTime = jobOrder.StartTime,
            EndTime = jobOrder.EndTime,
            EstimatedCompletionTime = jobOrder.EstimatedCompletionTime,
            TotalAmount = jobOrder.TotalAmount,
            DiscountAmount = jobOrder.DiscountAmount,
            TaxAmount = jobOrder.TaxAmount,
            GrandTotal = jobOrder.GrandTotal,
            Items = jobOrder.Items.Select(i => new JobOrderItemDto
            {
                Id = i.Id,
                JobOrderId = i.JobOrderId,
                ServiceId = i.ServiceId,
                ServiceName = i.ServiceName,
                Description = i.Description,
                Quantity = i.Quantity,
                UnitPrice = i.UnitPrice,
                Subtotal = i.Subtotal,
                Notes = i.Notes
            }).ToList(),
            Parts = jobOrder.Parts.Select(p => new JobOrderPartDto
            {
                Id = p.Id,
                JobOrderId = p.JobOrderId,
                ProductId = p.ProductId,
                ProductName = p.ProductName,
                ProductSku = p.ProductSku,
                Quantity = p.Quantity,
                UnitPrice = p.UnitPrice,
                Subtotal = p.Subtotal,
                Notes = p.Notes
            }).ToList(),
            CreatedAt = jobOrder.CreatedAt,
            UpdatedAt = jobOrder.UpdatedAt
        };
    }
}

File: src/Clients/talyerstudio-dashboard/src/services/jobOrderService.ts
import api, { API_URLS } from './api';

export interface JobOrderItem {
  id?: string;
  serviceId: string;
  serviceName?: string;
  description?: string;
  quantity: number;
  unitPrice: number;
  notes?: string;
}

export interface JobOrderPart {
  id?: string;
  productId: string;
  productName?: string;
  productSku?: string;
  quantity: number;
  unitPrice: number;
  notes?: string;
}

export interface JobOrder {
  id: string;
  jobOrderNumber: string;
  tenantId: string;
  branchId?: string;
  customerId: string;
  customerName?: string;
  vehicleId: string;
  vehicleInfo?: string;
  status: 'PENDING' | 'IN_PROGRESS' | 'COMPLETED' | 'INVOICED' | 'CANCELLED';
  priority: 'LOW' | 'NORMAL' | 'HIGH' | 'URGENT';  // CHANGED
  odometerReading?: number;
  customerComplaints?: string;
  inspectionNotes?: string;
  assignedMechanicIds?: string[];
  startTime?: string;
  endTime?: string;
  estimatedCompletionTime?: string;
  items: JobOrderItem[];
  parts: JobOrderPart[];
  totalAmount: number;
  discountAmount: number;
  taxAmount: number;
  grandTotal: number;
  createdAt: string;
  updatedAt?: string;
}

export interface CreateJobOrderRequest {
  customerId: string;
  vehicleId: string;
  priority: 'LOW' | 'NORMAL' | 'HIGH' | 'URGENT';  // CHANGED
  odometerReading?: number;
  customerComplaints?: string;
  inspectionNotes?: string;
  estimatedCompletionTime?: string;
  items: JobOrderItem[];
  parts: JobOrderPart[];
}
export interface UpdateJobOrderRequest extends CreateJobOrderRequest {
  id: string;
  status?: string;
}

const jobOrderService = {
  async getAll(): Promise<JobOrder[]> {
    const response = await api.get(`${API_URLS.JOBORDER}/job-orders`);
    return response.data;
  },

  async getById(id: string): Promise<JobOrder> {
    const response = await api.get(`${API_URLS.JOBORDER}/job-orders/${id}`);
    return response.data;
  },

  async getByCustomer(customerId: string): Promise<JobOrder[]> {
    const response = await api.get(`${API_URLS.JOBORDER}/job-orders/customer/${customerId}`);
    return response.data;
  },

  async getByStatus(status: string): Promise<JobOrder[]> {
    const response = await api.get(`${API_URLS.JOBORDER}/job-orders?status=${status}`);
    return response.data;
  },

  async create(data: CreateJobOrderRequest): Promise<JobOrder> {
    // Transform to match backend DTO exactly
    const payload = {
      customerId: data.customerId,
      vehicleId: data.vehicleId,
      priority: data.priority || 'NORMAL',
      odometerReading: data.odometerReading || 0,
      customerComplaints: data.customerComplaints || '',
      inspectionNotes: data.inspectionNotes || null,
      estimatedCompletionTime: data.estimatedCompletionTime || null,
      items: data.items.map(item => ({
        serviceId: item.serviceId,
        serviceName: item.serviceName || '',
        description: item.description || null,
        quantity: item.quantity,
        unitPrice: item.unitPrice,
        notes: item.notes || null
      })),
      parts: data.parts.map(part => ({
        productId: part.productId,
        productName: part.productName || '',
        productSku: part.productSku || null,
        quantity: part.quantity,
        unitPrice: part.unitPrice,
        notes: part.notes || null
      }))
    };

    console.log('Sending payload:', payload);
    
    const response = await api.post(`${API_URLS.JOBORDER}/job-orders`, payload);
    return response.data;
  },

  async update(id: string, data: UpdateJobOrderRequest): Promise<JobOrder> {
    const payload = {
      customerId: data.customerId,
      vehicleId: data.vehicleId,
      priority: data.priority || 'NORMAL',
      odometerReading: data.odometerReading || 0,
      customerComplaints: data.customerComplaints || '',
      inspectionNotes: data.inspectionNotes || null,
      estimatedCompletionTime: data.estimatedCompletionTime || null,
      items: data.items.map(item => ({
        serviceId: item.serviceId,
        serviceName: item.serviceName || '',
        description: item.description || null,
        quantity: item.quantity,
        unitPrice: item.unitPrice,
        notes: item.notes || null
      })),
      parts: data.parts.map(part => ({
        productId: part.productId,
        productName: part.productName || '',
        productSku: part.productSku || null,
        quantity: part.quantity,
        unitPrice: part.unitPrice,
        notes: part.notes || null
      }))
    };

    const response = await api.put(`${API_URLS.JOBORDER}/job-orders/${id}`, payload);
    return response.data;
  },

  async updateStatus(id: string, status: string): Promise<JobOrder> {
    const response = await api.patch(`${API_URLS.JOBORDER}/job-orders/${id}/status`, { status });
    return response.data;
  },

  async delete(id: string): Promise<void> {
    await api.delete(`${API_URLS.JOBORDER}/job-orders/${id}`);
  }
};

export default jobOrderService;